### DISCUSSION TITLE
Cypress: eq to the alias

### DISCUSSION THREAD

**Discussion by MaxAllianov** [cypress-io/cypress#15433] [Questions and Help] [answered]
Hi everyone!

I guess it is an easy question, but I'm stuck on this for an hour.
How can I count the table rows and match them with the text?

Here is the code: 
![image](https://user-images.githubusercontent.com/76516974/110812642-754fd600-8290-11eb-938b-ccd5c4cbe28a.png)

I have already found out how to count the TRs but I can't deal with the matching.
Here is my code at the moment:
`cy.get('#other_thai').parent().children('table').children('tbody').find('tr').its('length').should('eq',QUESTION)`

The code is already working if to set the constant numerical value (8 in my case)
I guess I should create an alias for the "QUESTION" using the regex (to get only numerical value from the text using the \d+) but I'm not sure how to do this.

Please advise how to create such alias
There is also a question: will it actually work? Can I use the alias for the eq?

Thank you for your help!

**Comment by bahmutov** [comment, marked as answer]
Cypress commands are asynchronous, so you cannot immediately use the alias or value, since it will be set only later. The solution is in https://glebbahmutov.com/cypress-examples/6.6.0/recipes/get-text-list.html#confirm-the-number-of-items and uses `.then` like this

```js
cy.get('#items-count')
  .invoke('text')
  .then(parseInt)
  .then((n) => {
    cy.get('#items li').should('have.length', n)
  })
```

**Reply by MaxAllianov** [reply]
Unfortunately, `parseInt ` returns 'NaN' because my text has the number value after the text.
Could you please also advise how to extract the number from the text?

**Reply by bahmutov** [reply]
Can you provide the html snippet with the number and the text? I'm sure it just requires a regex to grab the number. I would also suggest marking up the number in its own span element for simplicity

**Reply by MaxAllianov** [reply]
I've made a number pattern and match it with the text but for some reason, the code returns [ '8' ] against simple 8.
Here is a code I'm currently trying to make work:

```
cy.get('#other_thai').invoke('text')
            .then((text) => {
                const fullText = text;
                const pattern = /[0-9]/g;
                const number = fullText.match(pattern);
                cy.get('#other_thai').parent().children('table').children('tbody').find('tr').its('length').should('eq',number)
        })
```

And here is the issue:
![image](https://user-images.githubusercontent.com/76516974/111310690-0bf00e80-8666-11eb-9c2b-c83402fde93a.png)

Sorry I cannot share the html code cause it is under the internal company usage

**Reply by MaxAllianov** [reply]
```
<details>
<summary id="other_thai">Thailand (Other) [ 5 ]</summary>
<table>
<tbody>
    <tr></tr>
    <tr></tr>
    <tr></tr>
    <tr></tr>
    <tr></tr>
</tbody>
</table>
</details>
```

Here is a sample of a simple table. I'm trying to assert the tr count is matching the text in summary

```
cy.get('#other_thai').invoke('text')
            .then((text) => {
                const fullText = text;
                const pattern = new RegExp (/\d/g);
                const trNumber = fullText.match(pattern);
                cy.get('#other_thai').parent().children('table').children('tbody').find('tr')
                    .its('length').should('eq', trNumber)
```

The result will be "fail" because 8 doesn't match [ '8' ]
If to add `.toString()` function like this:
`const trNumber = fullText.match(pattern).toString();`
The result will be also "fail" but the reason is 8 doesn't match '8' 

I have tried to create a new variable and manually add the (') to the `.length` result, but it didn't worked. Seems like there is no way to remove the brackets (') using the [0-9]+g regex

**Reply by bahmutov** [reply]
Check your JavaScript (you can use `debugger` keyword to step through it). I posted the answer here https://glebbahmutov.com/cypress-examples/6.6.0/recipes/get-text-list.html#with-text-matching
```js
cy.get('#items-intro')
  .invoke('text')
  .then((s) => {
    // by adding an assertion here we print
    // the text in the command log for simple debugging
    expect(s).to.be.a('string')
    const matches = /\[\s*(\d+)\s*\]/.exec(s)
    return matches[1]
  })
  .then(parseInt)
  // another assertion to log the parsed number in the command log
  .should('be.a', 'number')
  // we expect between 1 and 10 items
  .and('be.within', 1, 10)
  .then((n) => {
    cy.get('#my-items li').should('have.length', n)
  })
```

**Reply by MaxAllianov** [reply]
Just figured out that the issue was in `its(length).should(eq, number)` and your answer appeared :D
For some reason `.its(length)` returns value of `" VALUE "` (with the quotes from both sides)
But `.should('have.length)'` returns the value without the quotes which allows to use it for matching.

Here is a final solution for my case that worked for my case:

```
cy.get('#other_thai').invoke('text')
            .then((text) => {
                const fullText = text;
                const pattern = new RegExp (/\d/g);
                const trNumber = fullText.match(pattern).toString();
                cy.log('trNumber', trNumber);
                cy.get('#other_thai').parent().children('table').children('tbody').find('tr').should('have.length',trNumber)      
                })
```

Thank you for your help!
